Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm

```diff
@@ -59,81 +61,67 @@
+static NSString * const documentStart = @"document_start";
+
 // FIXME: <https://webkit.org/b/261765> Consider adding support for cssOrigin.
 // FIXME: <https://webkit.org/b/261765> Consider adding support for injectImmediately.
+// FIXME: <https://webkit.org/b/264829> Add support for matchOriginAsFallback.
 
 namespace WebKit {
 
-static inline void parseTargetInjectionOptions(NSDictionary *targetInfo, WebExtensionScriptInjectionParameters& parameters, NSString **outExceptionString)
+NSArray *toWebAPI(const Vector<WebExtensionScriptInjectionResultParameters>& parametersVector, bool returnExecutionResultOnly)
 {
-    NSNumber *tabID = targetInfo[tabIDKey];
-    std::optional<WebExtensionTabIdentifier> identifier = toWebExtensionTabIdentifier(tabID.doubleValue);
-    if (!identifier) {
-        *outExceptionString = toErrorString(nil, tabIDKey, @"'%@' is not a tab identifier", tabID);
-        return;
-    }
-
-    parameters.tabIdentifier = identifier;
-
-    if (NSArray *frameIDs = targetInfo[frameIDsKey]) {
-        Vector<WebExtensionFrameIdentifier> frames;
-        for (NSNumber *frameID in frameIDs) {
-            auto identifier = toWebExtensionFrameIdentifier(frameID.doubleValue);
-            if (!isValid(identifier)) {
-                *outExceptionString = toErrorString(nil, frameIDsKey, @"'%@' is not a frame identifier", frameID);
-                return;
-            }
+    auto *results = [NSMutableArray arrayWithCapacity:parametersVector.size()];
 
-            frames.append(identifier.value());
+    // tabs.executeScript() only returns an array of the injection result.
+    if (returnExecutionResultOnly) {
+        for (auto& parameters : parametersVector) {
+            if (parameters.result)
+                [results addObject:parameters.result.value()];
+            else
+                [results addObject:NSNull.null];
         }
 
-        parameters.frameIDs = frames;
-        return;
+        return [results copy];
     }
 
-    if (!boolForKey(targetInfo, allFramesKey, false))
-        parameters.frameIDs = Vector { WebExtensionFrameConstants::MainFrameIdentifier };
-}
-
-static inline void parseScriptInjectionOptions(NSDictionary *script, WebExtensionScriptInjectionParameters& parameters)
-{
-    if (script[funcKey] || script[functionKey]) {
-        NSString *key = script[funcKey] ? funcKey : functionKey;
-        parameters.function = script[key];
-    }
-
-    if (script[argsKey] || script[argumentsKey]) {
-        NSString *key = script[argsKey] ? argsKey : argumentsKey;
-        parameters.arguments = makeVector<String>(script[key]);
-    }
+    for (auto& parameters : parametersVector) {
+        auto *result = [NSMutableDictionary dictionaryWithCapacity:3];
 
-    if (NSArray *files = script[filesKey])
-        parameters.files = makeVector<String>(files);
+        if (parameters.result)
+            result[@"result"] = (NSString *)parameters.result.value();
```

Comment:
I don't know if the cast is needed. I wasn't seeing error when doing this without the cast in one of @b-weinstein's PRs, so it might Just Work™.

Reply:
yea i don't think it's needed. i added it based on the feedback from the last PR but will revert back

Reply:
Yeah, I have thought it was needed in this case but if it works without, great!